subprojects/core/src/main/java/org/gradle/api/plugins/ExtensionContainer.java

```diff
@@ -21,28 +21,76 @@
 import org.gradle.api.UnknownDomainObjectException;
 import org.gradle.internal.HasInternalProtocol;
 
+import java.util.Map;
+
 /**
  * Allows adding 'namespaced' DSL extensions to a target object.
  */
 @HasInternalProtocol
 public interface ExtensionContainer {
 
     /**
+     * Adds a new extension to this container.
+     *
      * Adding an extension of name 'foo' will:
      * <li> add 'foo' dynamic property
      * <li> add 'foo' dynamic method that accepts a closure that is a configuration script block
      *
-     * @param name Will be used as a sort of namespace of properties/methods.
-     * @param extension Any object whose methods and properties will extend the target object
+     * The extension will be exposed as {@code publicType}.
+     *
+     * @param name The name for the extension
+     * @param publicType The extension public type
+     * @param extension Any object implementing {@code publicType}
+     * @param <P> the extension public type
+     * @param <I> the extension instance type
      * @throws IllegalArgumentException When an extension with the given name already exists.
+     * @since 4.0
+     */
+    @Incubating
+    <P, I extends P> void add(String name, Class<P> publicType, I extension);
```

Comment:
We could use `Object` instead.
But then we would allow consumers of the API to add arbitrary objects as extensions, possibly not implementing the public type. In this case, errors will pop up at runtime which I think is not nice. Plus we'll have to rely on doggy casts in our implementation of `ExtensionContainer` and `ExtensionStorage`.